docs/source/cpp/streaming_execution.rst

```diff
@@ -175,9 +175,607 @@ their completion::
+This is used to find the index of a particular scalar value. 
+
+Example::
+
+    arrow::compute::IndexOptions index_options(arrow::MakeScalar("1"));
+
+An example for creating an aggregate node::
+
+    arrow::compute::CountOptions options(arrow::compute::CountOptions::ONLY_VALID);
+
+    auto aggregate_options = arrow::compute::AggregateNodeOptions{
+        /*aggregates=*/{{"hash_count", &options}},
+        /*targets=*/{"a"},
+        /*names=*/{"count(a)"},
+        /*keys=*/{"b"}};
+
+    ARROW_ASSIGN_OR_RAISE(cp::ExecNode * aggregate,
+                              cp::MakeExecNode("aggregate", plan.get(), {source},
+                              aggregate_options));
+
+
+Scan-Node
+---------
+
+There is no class or struct defined as ScanNode in the source. 
+But :class:`arrow::compute::ScanNodeOptions` container includes the options
+passed to `MakeScanNode` internal function which creates an :class:`ExecNode`
+performing the defined task. This component includes a few options,
+defined in the :class:`arrow::compute::ScanNodeOptions` and this requires, 
+`std::shared_ptr<arrow::dataset::Dataset>`, 
+`std::shared_ptr<arrow::compute::ScanOptions>`, 
+`std::shared_ptr<arrow::util::AsyncToggle>`. 
+
+The :class:`arrow::compute::ScanOptions` includes the scaning options::
+
+    arrow::compute::Expression Materialize(std::vector<std::string> names,
+                              bool include_aug_fields = false) {
+        if (include_aug_fields) {
+            for (auto aug_name : {"__fragment_index",
+                "__batch_index", "__last_in_fragment"}) {
+            names.emplace_back(aug_name);
+            }
+        }
+
+        std::vector<arrow::compute::Expression> exprs;
+        for (const auto& name : names) {
+            exprs.push_back(arrow::compute::field_ref(name));
+        }
+
+        return arrow::compute::project(exprs, names);
+    }
+
+    auto options = std::make_shared<arrow::dataset::ScanOptions>();
+    // sync scanning is not supported by ScanNode
+    options->use_async = true;
+    options->projection = Materialize({});  // create empty projection
+
+    // construct the scan node
+    cp::ExecNode* scan;
+    auto scan_node_options = arrow::dataset::ScanNodeOptions{
+                              /*dataset*/dataset, 
+                              /*scan_options*/options};
+
+    ARROW_ASSIGN_OR_RAISE(scan,
+                            arrow::compute::MakeExecNode("scan", 
+                              plan.get(), 
+                              {}, 
+                              scan_node_options));
+
+``SinkNode``
+------------
+
+:class:`SinkNode` can be considered as the output or final node of an streaming 
+execution definition. :class:`arrow::compute::SinkNodeOptions` interface is used to pass 
+the required options. Requires 
+`std::function<arrow::Future<arrow::util::optional<arrow::compute::ExecBatch>>()>* generator`
```

Comment:
Note that this has a user-friendly alias, `function<Future<T>()>` can be written as `arrow::AsyncGenerator<T>`

Reply:
Ah, but I don't know if `AsyncGenerator` is an external alias.  I could go either way.